src/main/java/org/datadog/jmxfetch/Instance.java

```diff
@@ -506,176 +600,252 @@ public boolean timeToCollect() {
+                } catch (IOException e) {
+                    throw e;
+                } catch (Exception e) {
+                    log.warn("Cannot get bean attributes or class name: " + e.getMessage(), e);
+                    continue;
+                }
+
+                addMatchingAttributesForBean(beanName, className, attributeInfos);
+            }
+            log.info("Found {} matching attributes, collecting {} metrics total", matchingAttributes.size(), this.metricCountForMatchingAttributes);
+        } finally {
+            this.beanAndAttributeLock.unlock();
+        }
+    }
+
+    public void beanRegistered(ObjectName beanName) {
+        this.beanAndAttributeLock.lock();
+        log.debug("Bean registered event. {}", beanName);
+        String className;
+        MBeanAttributeInfo[] attributeInfos;
+        try {
+            log.debug("Getting class name for bean: " + beanName);
+            className = connection.getClassNameForBean(beanName);
+
+            // Get all the attributes for bean_name
+            log.debug("Getting attributes for bean: " + beanName);
+            attributeInfos = connection.getAttributesForBean(beanName);
+
+            this.addMatchingAttributesForBean(beanName, className, attributeInfos);
+            this.beans.add(beanName);
+        } catch (IOException e) {
+            // Nothing to do, connection issue
+            log.warn("Could not connect to get bean attributes or class name: " + e.getMessage());
+        } catch (Exception e) {
+            log.warn("Cannot get bean attributes or class name: " + e.getMessage(), e);
+        } finally {
+            log.debug("Bean registration processed. {}", beanName);
+            this.beanAndAttributeLock.unlock();
+        }
+    }
+    public void beanUnregistered(ObjectName mBeanName) {
+        this.beanAndAttributeLock.lock();
+        try {
+            int removedMetrics = 0;
+            int removedAttributes = 0;
+            for (Iterator<JmxAttribute> it = this.matchingAttributes.iterator(); it.hasNext();) {
+                JmxAttribute current = it.next();
+                if (current.getBeanName().compareTo(mBeanName) == 0) {
+                    it.remove();
+                    removedMetrics += current.getLastMetricsCount();
+                    removedAttributes++;
                 }
             }
+            log.info("Bean unregistered event, removed {} attributes ({} metrics) matching bean {}", removedAttributes, removedMetrics, mBeanName);
+            this.metricCountForMatchingAttributes -= removedMetrics;
+        } finally {
+            log.debug("Bean unregistration processed. {}", mBeanName);
+            this.beanAndAttributeLock.unlock();
+        }
+    }
+
+
+    private void subscribeToBeans() {
+        List<String> beanScopes = this.getBeansScopes();
+        BeanSubscriber subscriber = new BeanSubscriber(beanScopes, this.connection, this);
```

Comment:
This may be more complicated than needed, in my testing I see that JMX MBean notifications come in from a `ClientNotif` thread that seems to be managed by jmx, my thought here was that the thread which subscribed to the notifications would be the one notified, but I don't think that assumption is correct.